File: solutions/c/dnd-character/3/dnd_character.c
#include "dnd_character.h"

#include <stdlib.h>
#include <time.h>
#include <math.h>

static int dice_roll() {
    srand((unsigned int)time(NULL));
    return rand() % 6 + 1;
}

int ability() {
    int mini = dice_roll();
    int roll = 0;
    int sum = 0;
    for (unsigned int i = 0; i < 3; ++i) {
        roll = dice_roll();
        if (roll < mini) {
            sum += mini;
            mini = roll;
        } else {
            sum += roll;
        }
    }
    return sum;
}

int modifier(const int score) {
    return floor((score - 10) / 2.0);
}

dnd_character_t make_dnd_character() {
    dnd_character_t chr = {
        .strength = ability(),
        .dexterity = ability(),
        .constitution = ability(),
        .intelligence = ability(),
        .wisdom = ability(),
        .charisma = ability(),
    };
    chr.hitpoints = modifier(chr.constitution) + 10;
    return chr;
}

File: solutions/go/dnd-character/3/dnd_character.go
package dndcharacter

import (
    "math"
    "math/rand/v2"
)

type Character struct {
	Strength, Dexterity, Constitution,
    Intelligence, Wisdom, Charisma, Hitpoints int
}

func Modifier(score int) int {
	return int(math.Floor(float64(score-10) / 2))
}

func Ability() (sum int) {
	min := DiceRoll()
    for i := 0; i < 3; i++ {
        d := DiceRoll()
        if min <= d {
            sum += d
        } else {
            min, sum = d, sum + min
        }
    }
    return
}

func Abilities() (str, dex, con, itl, wis, chr int) {
    return Ability(), Ability(), Ability(),
    	Ability(), Ability(), Ability()
}

func GenerateCharacter() Character {
    str, dex, con, itl, wis, chr := Abilities()
	return Character{
        str, dex, con, itl,
        wis, chr, 10 + Modifier(con),
    }
}

func DiceRoll() int {
    return rand.N(6) + 1
}
